From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoftikLord <dimap9986@gmail.com>
Date: Thu, 6 Apr 2023 15:42:33 +0500
Subject: [PATCH] Experemetnal-Shulker-Optimize


diff --git a/src/main/java/net/minecraft/server/TileEntityShulkerBox.java b/src/main/java/net/minecraft/server/TileEntityShulkerBox.java
index 443fa5d63d254e785f3d80c5ba6cfda833107563..20781ab47fecb56cdf9f451b825746d0cf373424 100644
--- a/src/main/java/net/minecraft/server/TileEntityShulkerBox.java
+++ b/src/main/java/net/minecraft/server/TileEntityShulkerBox.java
@@ -235,7 +235,8 @@ public class TileEntityShulkerBox extends TileEntityLootable implements ITickabl
         dataconvertermanager.a(DataConverterTypes.BLOCK_ENTITY, (DataInspector) (new DataInspectorItemList(TileEntityShulkerBox.class, new String[] { "Items"})));
     }
 
-    public void load(NBTTagCompound nbttagcompound) {
+    public void load(NBTTagCompound nbttagcompound, EntityHuman entityhuman) {
+        if (entityhuman == null) return; // Lamius - Shulker Optimize
         super.load(nbttagcompound);
         this.e(nbttagcompound);
     }
