From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoftikLord <dimap9986@gmail.com>
Date: Thu, 6 Apr 2023 16:21:36 +0500
Subject: [PATCH] New-Test-patch


diff --git a/src/main/java/net/minecraft/server/TileEntityChest.java b/src/main/java/net/minecraft/server/TileEntityChest.java
index 85cdd87a73137103643f09f5493de3506807e497..31a0242d877b5758aba025f5885f369fc4cf3b5e 100644
--- a/src/main/java/net/minecraft/server/TileEntityChest.java
+++ b/src/main/java/net/minecraft/server/TileEntityChest.java
@@ -83,21 +83,6 @@ public class TileEntityChest extends TileEntityLootable { // Paper - Remove ITic
     public static void a(DataConverterManager dataconvertermanager) {
         dataconvertermanager.a(DataConverterTypes.BLOCK_ENTITY, (DataInspector) (new DataInspectorItemList(TileEntityChest.class, new String[] { "Items"})));
     }
-
-    public void load(NBTTagCompound nbttagcompound, EntityHuman entityhuman) {
-        if (entityhuman == null) return; // Lamius - Shulker Optimize
-        super.load(nbttagcompound);
-        this.items = NonNullList.a(this.getSize(), ItemStack.a);
-        if (!this.c(nbttagcompound)) {
-            ContainerUtil.b(nbttagcompound, this.items);
-        }
-
-        if (nbttagcompound.hasKeyOfType("CustomName", 8)) {
-            this.o = nbttagcompound.getString("CustomName");
-        }
-
-    }
-
     public NBTTagCompound save(NBTTagCompound nbttagcompound) {
         super.save(nbttagcompound);
         if (!this.d(nbttagcompound)) {
@@ -286,7 +271,16 @@ public class TileEntityChest extends TileEntityLootable { // Paper - Remove ITic
         }
     }
 
-    public void startOpen(EntityHuman entityhuman) {
+    public void startOpen(EntityHuman entityhuman, NBTTagCompound nbttagcompound) {
+        super.load(nbttagcompound);
+        this.items = NonNullList.a(this.getSize(), ItemStack.a);
+        if (!this.c(nbttagcompound)) {
+            ContainerUtil.b(nbttagcompound, this.items);
+        }
+
+        if (nbttagcompound.hasKeyOfType("CustomName", 8)) {
+            this.o = nbttagcompound.getString("CustomName");
+        }
         if (!entityhuman.isSpectator()) {
             if (this.l < 0) {
                 this.l = 0;
