From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Fri, 3 Jun 2022 02:58:12 +0500
Subject: [PATCH] feat: add TileEntityTickEvent


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 72b23385bc5ec9bf730217e0e591eb82435425a5..c5446704642d26341d512a52510fd8cda17e270d 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -127,6 +127,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
     // Lamius start
     boolean hasEntityMoveEvent;
     boolean hasEntityTickEvent;
+    boolean hasTileEntityTickEvent;
     // Lamius end
 
     public MinecraftServer(OptionSet options, Proxy proxy, DataConverterManager dataconvertermanager, YggdrasilAuthenticationService yggdrasilauthenticationservice, MinecraftSessionService minecraftsessionservice, GameProfileRepository gameprofilerepository, UserCache usercache) {
@@ -889,6 +890,7 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
         // Lamius start
         hasEntityMoveEvent = me.core.lamius.Lamius.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length != 0;
         hasEntityTickEvent = me.core.lamius.Lamius.event.entity.EntityTickEvent.getHandlerList().getRegisteredListeners().length != 0;
+        hasTileEntityTickEvent = me.core.lamius.Lamius.event.block.TileEntityTickEvent.getHandlerList().getRegisteredListeners().length != 0;
         // Lamius end
 
         int i;
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 69ceb76bd9b198d4aeafea134608cb04e3efab23..cab88fa55e3790fe427d75cba05e8fa06aef4c8b 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1706,6 +1706,7 @@ public abstract class World implements IBlockAccess {
 
         this.O = true;
         // Spigot start
+        boolean hasTileEntityTickEvent = getMinecraftServer().hasTileEntityTickEvent; // Lamius - Add TileEntityTickEvent
         // Iterator iterator = this.tileEntityListTick.iterator();
         int tilesThisCycle = 0;
         for (tileTickPosition = 0; tileTickPosition < tileEntityListTick.size(); tileTickPosition++) { // Paper - Disable tick limiters
@@ -1730,6 +1731,7 @@ public abstract class World implements IBlockAccess {
                     shouldTick = shouldTick && !chunk.isUnloading() && chunk.scheduledForUnload == null;
                 if (shouldTick && this.P.a(blockposition)) {
                     // Paper end
+                    if (!hasTileEntityTickEvent || CraftEventFactory.callTileEntityTickEvent(tileentity)) // Lamius - Add TileEntityTickEvent
                     try {
                         /* // Lamius start - Remove MethodProfiler
                         this.methodProfiler.a(() -> {
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 97331a2179aaba64c232992e14ae8ddbb3fff14c..c04504623c38bc77ea7589f33020c33bcc112537 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -77,6 +77,7 @@ public class CraftEventFactory {
     private static final me.core.lamius.Lamius.event.entity.EntityMoveEvent entityMoveEvent = new me.core.lamius.Lamius.event.entity.EntityMoveEvent(null);
     private static final me.core.lamius.Lamius.event.entity.AnyEntitySpawnEvent anyEntitySpawnEvent = new me.core.lamius.Lamius.event.entity.AnyEntitySpawnEvent(null);
     private static final me.core.lamius.Lamius.event.entity.EntityTickEvent entityTickEvent = new me.core.lamius.Lamius.event.entity.EntityTickEvent(null);
+    private static final me.core.lamius.Lamius.event.block.TileEntityTickEvent tileEntityTickEvent = new me.core.lamius.Lamius.event.block.TileEntityTickEvent(null);
     // Lamius end
 
     // helper methods
@@ -1216,5 +1217,18 @@ public class CraftEventFactory {
 
         return !event.isCancelled();
     }
+
+    public static boolean callTileEntityTickEvent(TileEntity tileEntity) {
+        me.core.lamius.Lamius.event.block.TileEntityTickEvent event = CraftEventFactory.tileEntityTickEvent;
+
+        BlockPosition pos = tileEntity.getPosition();
+        Block block = tileEntity.getWorld().getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+
+        event.setData(block);
+
+        Bukkit.getPluginManager().callEvent(event);
+
+        return !event.isCancelled();
+    }
     // Lamius end
 }
