From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoftikLord <dimap9986@gmail.com>
Date: Tue, 28 Mar 2023 22:16:34 +0500
Subject: [PATCH] Fix-Packet-Fly


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index dc4baee9552c22682f1e63ab14e13424b51674bc..371d27d5668aed30c27932a7e5f7aacceabe7690 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -135,6 +135,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
     private float lastYaw = Float.MAX_VALUE;
     private boolean justTeleported = false;
     private boolean hasMoved; // Spigot
+    private boolean lagbackTeleport; // Lamius
 
     public CraftPlayer getPlayer() {
         return (this.player == null) ? null : (CraftPlayer) this.player.getBukkitEntity();
@@ -519,6 +520,15 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
 
     public void a(PacketPlayInTeleportAccept packetplayinteleportaccept) {
         PlayerConnectionUtils.ensureMainThread(packetplayinteleportaccept, this, this.player.x());
+        // Lamius start
+        if (packetplayinteleportaccept.a() == this.teleportAwait && this.lagbackTeleport) {
+            if (++this.teleportAwait == Integer.MAX_VALUE) {
+                this.teleportAwait = 0;
+            }
+            this.lagbackTeleport = false;
+            return;
+        }
+        // Lamius end
         if (packetplayinteleportaccept.a() == this.teleportAwait && this.teleportPos != null) { // CraftBukkit
             this.player.setLocation(this.teleportPos.x, this.teleportPos.y, this.teleportPos.z, this.player.yaw, this.player.pitch);
             if (this.player.L()) {
@@ -648,7 +658,11 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                                 if (d11 - d10 > Math.max(f2, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) /*&& (!this.minecraftServer.R() || !this.minecraftServer.Q().equals(this.player.getName()))*/) { // Spigot // Lamius - Remove singleplayer code
                                 // CraftBukkit end
                                     PlayerConnection.LOGGER.warn("{} moved too quickly! {},{},{}", this.player.getName(), Double.valueOf(d7), Double.valueOf(d8), Double.valueOf(d9));
-                                    this.a(this.player.locX, this.player.locY, this.player.locZ, this.player.yaw, this.player.pitch);
+                                    // Lamius start
+                                    // this.a(this.player.locX, this.player.locY, this.player.locZ, this.player.yaw, this.player.pitch);
+                                    sendPacket(new PacketPlayOutPosition(this.player.locX, this.player.locY, this.player.locZ, this.player.yaw, this.player.pitch, Collections.emptySet(), teleportAwait));
+                                    this.lagbackTeleport = true;
+                                    // Lamius end
                                     return;
                                 }
                             }
