From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoftikLord <dimap9986@gmail.com>
Date: Mon, 3 Apr 2023 14:53:23 +0500
Subject: [PATCH] Unmount-vehicle-upon-exit


diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index b8b4d12b4c2fe5dfaafe9346042fb685f1d59d1d..570aea795e9af387b50c449ee22f8b4cdae991ea 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -486,6 +486,7 @@ public abstract class PlayerList {
     }
 
     public String disconnect(EntityPlayer entityplayer) { // CraftBukkit - return string
+        entityplayer.stopRiding(); // Lamius - Unmount vehicle upon exit
         WorldServer worldserver = entityplayer.x();
 
         entityplayer.b(StatisticList.f);
